novelist.py

```diff
@@ -0,0 +1,309 @@
+import json
+import logging
+import urllib
+from nose.tools import set_trace
+
+from core.config import Configuration
+from core.coverage import (
+    CoverageFailure,
+    CoverageProvider,
+)
+from core.metadata_layer import (
+    ContributorData,
+    IdentifierData,
+    LinkData,
+    MeasurementData,
+    Metadata,
+    SubjectData,
+)
+from core.model import (
+    DataSource,
+    Hyperlink,
+    Identifier,
+    Measurement,
+    Representation,
+    Subject,
+)
+from core.util import TitleProcessor
+
+class NoveListAPI(object):
+
+    log = logging.getLogger("NoveList API")
+    version = "2.2"
+
+    # While the NoveList API doesn't require parameters to be passed via URL,
+    # the Representation object needs a unique URL to return the proper data
+    # from the database.
+    QUERY_ENDPOINT = "http://novselect.ebscohost.com/Data/ContentByQuery?\
+            ISBN=%(ISBN)s&ClientIdentifier=%(ClientIdentifier)s&version=%(version)s"
+    MAX_REPRESENTATION_AGE = 6*30*24*60*60      # six months
+
+    @classmethod
+    def from_config(cls, _db):
+        config = Configuration.integration(Configuration.NOVELIST_INTEGRATION)
+        profile = config.get(Configuration.NOVELIST_PROFILE)
+        password = config.get(Configuration.NOVELIST_PASSWORD)
+        if not (profile and password):
+            raise ValueError("No NoveList client configured.")
+        return cls(_db, profile, password)
+
+    def __init__(self, _db, profile, password):
+        self._db = _db
+        self.profile = profile
+        self.password = password
+        self.source = DataSource.lookup(self._db, DataSource.NOVELIST)
+
+    def lookup_equivalent_isbns(self, identifier):
+        """Finds NoveList data for all ISBNs equivalent to an identifier.
+
+        :return: a list of Metadata objects
+        """
+
+        license_source = DataSource.license_source_for(self._db, identifier)
+        # Look up strong ISBN equivalents.
+        lookup_metadata =  [self.lookup(eq.output)
+                for eq in identifier.equivalencies
+                if (eq.data_source==source and eq.strength==1
+                    and eq.output.type==Identifier.ISBN)]
+
+        if not lookup_metadata:
+            self.log.error("Identifiers without an ISBN equivalent can't \
+                    be looked up with NoveList: %r" % identifier)
+            return None
+
+        return [metadata for metadata in lookup_metadata if metadata]
+
+    def lookup(self, identifier):
+        """Requests NoveList metadata for a particular identifier
+
+        :return: None, a Metadata object, or a list of Metadata objects
```

Comment:
Actually, looking further down the code I see you say that you expect to only get one piece of data here -- all ISBNs for a literary work will return the same information, and you throw an exception if the ISBNs turn out to represent more than one literary work.

So I would expect this to either stop processing ISBNs as soon as it finds some data, or to gather data for all the ISBNs and choose the most commonly occuring item (reducing the chance that a single bad ISBN ruins everything).

In either case this method would return only one piece of information.